### PR TITLE
Standing PR for TensorFlow learning refactor

### PR DESCRIPTION
**Pull request by henryre**
* Refactor of learning primitives (`NoiseAwareModel`)
* Wrapper for TensorFlow models with automatic saving and loading utilities (`TFNoiseAwareModel`)
* Adds `LogisticRegression` as a supported model (updates tutorial to reflect)
* Updates to hyperparameter searchers
* Adds `reLSTM` as a contributor model for relation extraction

### CODE REVIEW

**Comment by ajratner** [comment]
Looking good so far... ping me on slack when ready for review!

**Comment by henryre** [comment]
@ajratner Ok good for a review!
cc: @stephenbach 

There are two major improvement opportunities, but they can come after the release:

* Sparse tensor computations for `LogisticRegression`
* Same-session model loading for `reLSTM`

**Comment by ajratner** [comment]
Ok I'll review a bit later tonight!

File: snorkel/contrib/learning/relstm.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        return x
+
+    def _preprocess_data(self, candidates, extend=False):
+        """Convert candidate sentences to lookup sequences"""
+        sentences = []
+        for c in candidates:
+            # Get arguments and lemma sequence
+            args = [
+                (c[0].get_word_start(), c[0].get_word_end(), 1),
+                (c[1].get_word_start(), c[1].get_word_end(), 2)
```
Comment (ajratner) [inline snorkel/contrib/learning/relstm.py]: Marker indices should be consistent with doc string above (1,2 vs. 0,1)

File: snorkel/contrib/learning/relstm.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            ]
+            s = self._mark_sentence(
+                [w.lower() for w in c.get_parent().lemmas], args
```
Comment (ajratner) [inline snorkel/contrib/learning/relstm.py]: Easy addition to make the sentence token type configurable here?

File: snorkel/contrib/learning/relstm.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            sentences.append(np.array([retriever(w) for w in s]))
+        return sentences
+
+    def _make_tensor(self, x):
+        """Construct input tensor with padding"""
+        batch_size = len(x)
+        tx = np.zeros((self.mx_len, batch_size), dtype=np.int32)
+        tlen = np.zeros(batch_size, dtype=np.int32)
+        # Pad or trim each x
+        # TODO: fix for arguments outside max length
```
Comment (ajratner) [inline snorkel/contrib/learning/relstm.py]: Either fix or explain what's going on here / what is currently implemented pre-fix

File: snorkel/contrib/learning/relstm.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            initial_state = cell.zero_state(batch_size, tf.float32)
+            rnn_outputs, _ = tf.nn.dynamic_rnn(
+                cell, inputs, initial_state=initial_state, time_major=True
+            )
+            # Get LSTM variables
+            self.save_dict = {
+                x.name: x for x in tf.get_collection(
+                tf.GraphKeys.GLOBAL_VARIABLES, scope=scope.name)
+            }
+        # Take mean across sentences
```
Comment (ajratner) [inline snorkel/contrib/learning/relstm.py]: Is it standard to take mean across sentences here?

File: snorkel/learning/disc_learning.py [hunk 5]
```diff
@@ -42,200 +37,85 @@ def score(self, session, X_test, test_labels, gold_candidate_set=None, b=0.5, se
         return s.score(test_marginals, train_marginals, b=b,
```
Comment (ajratner) [inline snorkel/learning/disc_learning.py]: Do we ever set `self.X_train` any more (see lines before, and again comes up below)?  If not we should get rid of this, and get the training marginals in a different way, or just not display them here?

Comment (ajratner) [inline snorkel/learning/disc_learning.py]: I'm fine just getting rid of this, whatever is fast & cleans this loose thread up

File: tutorials/intro/Intro_Tutorial_4.ipynb [hunk 6]
```diff
@@ -471,8 +478,8 @@
    },
    "outputs": [],
    "source": [
-    "from snorkel.learning import LogReg\n",
-    "disc_model = LogReg()"
+    "from snorkel.learning import LogisticRegression\n",
+    "disc_model = LogisticRegression()"
```
Comment (ajratner) [inline tutorials/intro/Intro_Tutorial_4.ipynb]: Let's add a comment in the tutorial just mentioning the `reLSTM`?

File: snorkel/learning/disc_learning.py [hunk 7]
```diff
@@ -1,24 +1,19 @@
...
 class NoiseAwareModel(object):
     """Simple abstract base class for a model."""
-    def __init__(self, bias_term=False):
-        self.w         = None
-        self.bias_term = bias_term
-        self.X_train   = None
+
+    def __init__(self, name):
+        self.name = name
+        super(NoiseAwareModel, self).__init__()
```
Comment (henryre) [inline snorkel/learning/disc_learning.py]: Looks cool. Probably none though...

File: snorkel/learning/disc_learning.py (hunk 5 above)
Comment (henryre) [inline snorkel/learning/disc_learning.py]: I'll get rid of this. Never really display them in practice anyway.

File: snorkel/contrib/learning/relstm.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+from snorkel.learning import get_train_idxs, TFNoiseAwareModel
+from time import time
+
+
+class reLSTM(TFNoiseAwareModel):
+
+    def __init__(self, save_file=None, name='reLSTM'):
+        """LSTM for relation extraction"""
+        # Define metadata
+        self.mx_len          = None
```
Comment (henryre) [inline snorkel/contrib/learning/relstm.py]: 👍

File: snorkel/contrib/learning/relstm.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+        self.lr              = None
+        self.dropout         = None
+        # Define input layers
+        self.sentences       = None
+        self.sentence_length = None
+        self.y               = None
+        # Super constructor
+        super(reLSTM, self).__init__(save_file=save_file, name=name)
+
+    def _mark(self, l, h, idx):
```
Comment (henryre) [inline snorkel/contrib/learning/relstm.py]: 👍

File: snorkel/contrib/learning/relstm.py (hunk 1 above)
Comment (henryre) [inline snorkel/contrib/learning/relstm.py]: 🤓

File: snorkel/contrib/learning/relstm.py (hunk 2 above)
Comment (henryre) [inline snorkel/contrib/learning/relstm.py]: 👍

File: snorkel/contrib/learning/relstm.py (hunk 3 above)
Comment (henryre) [inline snorkel/contrib/learning/relstm.py]: Will add doc

File: snorkel/contrib/learning/relstm.py (hunk 4 above)
Comment (henryre) [inline snorkel/contrib/learning/relstm.py]: Yeah, simplest sentence embedding technique. Can improve down the road, but easy enough for now.

File: tutorials/intro/Intro_Tutorial_4.ipynb (hunk 6 above)
Comment (henryre) [inline tutorials/intro/Intro_Tutorial_4.ipynb]: Let's add it to another tutorial. LSTM will likely perform much worse on the intro tutorial than LR.

Comment (ajratner) [inline tutorials/intro/Intro_Tutorial_4.ipynb]: Ok if that's the case makes sense

**Comment by henryre** [comment]
@ajratner Changes made, testing out now

**Comment by henryre** [comment]
No even one invited you, Coveralls #yolo

**Pull request merged**